### ISSUE TITLE
WebAssembly Port

### ISSUE THREAD

**Issue by lalalune** [mapeditor/tiled#3554]
Heya, this is a pretty out-there question, but as a wasm port been considered? I'm sure file I/O may need to be considered, but other than, it seems like Qt can build to wasm. Web-based editor would be very handy, and could also see integration or embedding into games and such :)

**Comment by bjorn** [comment]
Yes, it's been considered and I've also compiled Tiled against Qt for WebAssembly and ran it in my browser. However, it will be very hard to turn it into a usable editor when running in that environment.

The main problem is the lack of direct file access. Qt provides functions to read/write single user-selected files, but Tiled generally wants to open a lot of other files, like the referenced tileset or image files. Or JS extensions, when opening a project. I'm not sure if in the meantime it would be possible to grant access to an entire folder, which would mostly resolve this issue.

Alternatively to Tiled in the browser accessing a local folder would of course be to enable accessing files on Dropbox or other services, though that will require a lot of custom integration code.

A more minor issue, but one that would take a lot of effort to improve, is that the UI isn't generally very suitable for running in a browser. On many occasions it wants to open popup windows and I'm not sure how well the dock widgets in its main window would be supported. Switching to a QtQuick based interface would be a possible solution to this problem, and would also make map rendering a lot faster by using the hardware, but it will be a major project.

**Comment by SaadBazaz** [comment]
Bump. I think the file system issue can be solved with stuff like [web-fs](https://github.com/newDINO/web-fs) (uses the [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API)).
UI-wise yes, we'd need to decouple the rendering engine from the views, and maybe then separate webviews can be made (built in native web technologies like React). I think even just separating the fileparser (which converts from/to .tiled files) can be a step in the right direction.

I wonder how far this would go. I do see very high demand for this for a lot of studios, can be a paid offering. A lot of companies (including my own) end up making their own internal tool for this kind of usecase, whereas Tiled is an industry standard.

cc @lalalune @bjorn
